src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -463,5 +463,32 @@ internal static bool IsVoidOrUnit(Type type)
 
             return type == typeof(void) || type.FullName == "Microsoft.FSharp.Core.Unit";
         }
+
+        /// <summary>
+        /// Returns the property information for a given named indexer.
+        /// </summary>
+        /// <param name="type">Type to reflect on for indexer.</param>
+        /// <param name="indexerTypes">List of indexer types that matches the indexer type order.</param>
+        /// <returns>The return value from the invoked method</returns>
+        public static PropertyInfo? GetIndexerByName(Type type, Type[] indexerTypes)
+        {
+            const string defaultFrameworkIndexerName = "Item";
```

Comment:
The default indexer name varies. It's `"Item"` when there is no DefaultMemberAttribute applied to the class.

You can write a test for this by putting `[IndexerName("NonDefaultIndexerName")]` on a C# indexer. The C# compiler removes this attribute during emit and emits `[DefaultMember("NonDefaultIndexerName")]` on the class instead.

([sharplab link](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAGAAkIEYA6AJQFcA7DbAWxioGEIWAHPGKAGUBAN2xgYAZwDcAWABQCwgCZSHBQG8FpHaQDaASQYATGAgEA5AIZsAFACILEBgBEYAMyt1cGI6fNQ1mz2AJQAutq6hADMpNhMpBgAFtiSevEYpAhhpAC8AHxZcvIAvkA))

Reply:
I had known about this, but with this comment:
https://github.com/nunit/nunit/issues/2211#issuecomment-305782721

I can't remember why I moved away from handling that condition. I'll have to look back into what was going on there. Let me check my reference test and get back on this.

Reply:
Oh, I'm sorry for the confusion in terms. https://github.com/nunit/nunit/issues/2211#issuecomment-305782721 refers to default versus named *indexers*. A default indexer looks like `obj[args]`. A named indexer is not a concept in C# but it would look like `obj.SomeProperty[args]`. In VB, the default indexer is declared with the keyword `Default` rather than being given a name. We decided to only implement assertions on default indexers.

My comment here in this PR is on a different, independent topic. This topic is the *name* of a default indexer. The default name of a default indexer is `Item`, but you can declare a *default* indexer with a *non-default* name. This isn't something we should just choose to ignore; it's an actual correctness issue that would lead to bad behavior if we ignore the DefaultMember attribute.

Reply:
I've added some more tests and implementation around the handling `[IndexerName("NonDefaultIndexerName")]`

I've changed the reflection code now to return the `MethodInfo` rather than the `PropertyInfo`. The method now attempts get default indexer first, otherwise it goes into a property search.